### ISSUE TITLE
-Woverflow error when compiling on ppc64le

### ISSUE THREAD

**Issue by nmeum** [ccache/ccache#820]
While working on the Alpine Linux Edge package for ccache (4.2). I noticed that it doesn't compile fine on ppc64le when file cloning support is enabled:

```
/builds/alpine/aports/main/ccache/src/ccache-4.2/src/Util.cpp: In function 'void Util::clone_file(const string&, const string&, bool)':
/builds/alpine/aports/main/ccache/src/ccache-4.2/src/Util.cpp:249:23: error: overflow in conversion from 'long unsigned int' to 'int' changes value from '2147783689' to '-2147183607' [-Werror=overflow]
  249 |   if (ioctl(*dest_fd, FICLONE, *src_fd) != 0) {
      |                       ^~~~~~~
cc1plus: all warnings being treated as errors
```

Not sure if that's actually an issue as the `FICLONE` macro is provided by the kernel headers (also not sure if there is any easy way to get rid of it). However, I also noticed the following compiler overflow warning regarding a memcpy invocation:

```
In function 'compress_parents_parallel',
    inlined from 'compress_subtree_to_parent_node' at /builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:352:9,
    inlined from 'blake3_hasher_update.part.0' at /builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:536:7:
/builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:240:5: warning: 'memcpy' writing 32 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
  240 |     memcpy(&out[parents_array_len * BLAKE3_OUT_LEN],
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  241 |            &child_chaining_values[2 * parents_array_len * BLAKE3_OUT_LEN],
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  242 |            BLAKE3_OUT_LEN);
      |            ~~~~~~~~~~~~~~~
```

I haven't investigated the latter warning in greater details. I also don't have direct access to the ppc64le architecture. I just figured that reporting it would be better than keeping this to myself. If you are not interested in bug reports regarding obscure architectures, just close this issue :)

**Comment by jrosdahl** [comment]
Thanks for the bug report.

> Not sure if that's actually an issue as the `FICLONE` macro is provided by the kernel headers (also not sure if there is any easy way to get rid of it).

Don't know either, but spontaneously it sounds like a libc/kernel compatibility issue. I suggest compiling without `-Werror`.

>However, I also noticed the following compiler overflow warning regarding a memcpy invocation:
> 
> ```
> In function 'compress_parents_parallel',
>     inlined from 'compress_subtree_to_parent_node' at /builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:352:9,
>     inlined from 'blake3_hasher_update.part.0' at /builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:536:7:
> /builds/alpine/aports/main/ccache/src/ccache-4.2/src/third_party/blake3/blake3.c:240:5: warning: 'memcpy' writing 32 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
>   240 |     memcpy(&out[parents_array_len * BLAKE3_OUT_LEN],
>       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   241 |            &child_chaining_values[2 * parents_array_len * BLAKE3_OUT_LEN],
>       |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>   242 |            BLAKE3_OUT_LEN);
>       |            ~~~~~~~~~~~~~~~
> ```

That originates from third party code that is bundled with ccache so I don't know about details here. Regardless, I don't understand how the compiler can get the impression that `&out[parents_array_len * BLAKE3_OUT_LEN]` points to a size 0 region when `out` is a `uint8_t*` function parameter.

> I haven't investigated the latter warning in greater details. I also don't have direct access to the ppc64le architecture. I just figured that reporting it would be better than keeping this to myself. If you are not interested in bug reports regarding obscure architectures, just close this issue :)

OK. I'll go ahead closing it, but it's great to have it searchable in case others notice the same issue. :slightly_smiling_face:

**Comment by MaxPeal** [comment]
"I believe this is fixed by b8e2dda, but please reopen if I've missed something." 
https://github.com/BLAKE3-team/BLAKE3/issues/183#issuecomment-904885973

**Comment by jrosdahl** [comment]
> "I believe this is fixed by b8e2dda, but please reopen if I've missed something." [BLAKE3-team/BLAKE3#183 (comment)](https://github.com/BLAKE3-team/BLAKE3/issues/183#issuecomment-904885973)

Thanks for noting this. BLAKE3 was upgraded in 10bfc8264bc7b3e161edc25d47a6e967fa85d349.
